### PR TITLE
feat: Allow disabling of editor templates when accepting method/function invocation completions (fixes #718)

### LINKED ISSUE #718: When code completions with a single tab stop are inserted, it should be possible to disable the resulting template
Currently when resolved code completions with tab stops are inserted, the tab stops are added to an editor template. That works well and is likely what most would want when there are multiple tab stops, but when there's only one, the single tab stop actually gets in the way of a smooth typing experience because it has to be finished with **Enter**, **Tab**, or **Esc**, and doing so moves the caret to another location. That also causes those keys to have an ambiguous meaning when typing a value for the tab stop that triggers auto-completion.

For example, typing **Tab** in the following both inserts the currently selected completion and finishes the editor template, moving the caret out of the parentheses:

Image

That's not what's wanted if the user is accepting that completion as part of a larger expression for the current invocation argument, and then the user has to return the caret to the desired location. That's what I meant by "gets in the way of a smooth typing experience".

Minimally it should be possible to disable this editor template behavior when there's a single tab stop in the inserted completion. Instead any inserted default argument text should be pre-selected so that it's overwritten if/when the user begins typing. That would likely take the form of a `completion` client configuration option (see #705 for an existing PR with an existing `completion` client configuration option).

### PR DESCRIPTION
**Pull request by SCWells72**
This is a fresh branch-based PR as a replacmeent for #738.

As stated/shown in #718, the editor template that is currently started when accepting a code completion that is a method/function invocation ~~with a single argument~~ is quite disruptive to the standard editor flow. ~~This PR adds a new client configuration option, `completion.useTemplateForSingleArgument`, that, when disabled (enabled by default for backward-compatibility), instead just places the caret into the invocation arguments so that the user can populate them as they would in any of JetBrains' natively-supported language editors. I've also updated all existing language server definition templates to disable editor templates in this case so that the "default" behavior for newly-created language server definitions is the smoother/more natural editor flow, but existing language server definitions will not see a change in behavior.~~

**UPDATE:** This has now been generalized to allow disabling editor templates for _any_ invocation-only snippet. An invocation-only snippet is one where all non-end (`$0`) variables are found as a comma-/whitespace-delimited list in invocation parentheses. If other invocation patterns are found in LSP languages, recognition support can be added for them as well.

The configuration option is now `completion.useTemplateForInvocationOnlySnippet` to reflect this more general behavior. As before, it is disabled by default in code but enabled by default in the bundled language server templates.

This also properly handles the situation of a single no-value variable snippet by forgoing an editor template and moving to the correct caret position for that variable with no need for any client config change.

### CODE REVIEW

File: src/main/java/com/redhat/devtools/lsp4ij/client/features/LSPCompletionProposal.java [hunk 1]
```diff
@@ -100,14 +114,16 @@ public void handleInsert(@NotNull InsertionContext context) {
         updateCompletionItemFromResolved();
         Template template = null;
         if (item.getInsertTextFormat() == InsertTextFormat.Snippet) {
+            // Adjust the snippet content as appropriate based on client config
+            String snippetContent = adjustSnippetContent(getInsertText());
```
Comment (SCWells72) [inline src/main/java/com/redhat/devtools/lsp4ij/client/features/LSPCompletionProposal.java:118]: Now we adjust the LSP-provided snippet before parsing it into a template if appropriate based on client config and the snippet contents.

File: src/main/java/com/redhat/devtools/lsp4ij/client/features/LSPCompletionProposal.java [hunk 2]
```diff
@@ -116,7 +132,12 @@ public void handleInsert(@NotNull InsertionContext context) {
         // Apply all text edits
         apply(context.getDocument(), context.getOffset(CompletionInitializationContext.SELECTION_END_OFFSET));
 
-        if (shouldStartTemplate(template)) {
+        // Just move to the first tab stop if there's a single invocation argument with no default value
+        if (shouldMoveToFirstTabStop(template)) {
```
Comment (SCWells72) [inline src/main/java/com/redhat/devtools/lsp4ij/client/features/LSPCompletionProposal.java:136]: Now we only move to the first tab stop for a template with a single non-end variable that has no value. That happens independent of client config because it doesn't make sense to start an editor template for a single tab stop with no default value. Instead just place the caret in that same location and let the user type what they need.

File: src/main/java/com/redhat/devtools/lsp4ij/client/features/LSPCompletionProposal.java [hunk 3]
```diff
@@ -155,12 +176,57 @@ private void updateCompletionItemFromResolved() {
         this.item = resolved;
     }
 
+    @NotNull
+    private String adjustSnippetContent(@NotNull String snippetContent) {
```
Comment (SCWells72) [inline src/main/java/com/redhat/devtools/lsp4ij/client/features/LSPCompletionProposal.java:180]: Hopefully this method is pretty self-explanatory, but basically if client config says not to use editor templates for invocation arg-only completion snippets, see if the snippet contains what appears to be an invocation arg list with template variables, and if so **and** if there are no non-end variables outside of that invocation arg list, just update the snippet to be `<beforeArgList>($0)<afterArgList>` with any other `$0` removed. That way when the snippet is parsed into a template, it will just result in the caret being moved into the invocation arg parens.

File: src/main/java/com/redhat/devtools/lsp4ij/client/features/LSPCompletionProposal.java [hunk 4]
```diff
@@ -155,12 +176,57 @@ private void updateCompletionItemFromResolved() {
...
+                String beforeInvocationArgs = snippetContent.substring(0, invocationArgsStartIndex + 1);
+                String afterInvocationArgs = snippetContent.substring(invocationArgsEndIndex - 1);
+                snippetContent = beforeInvocationArgs.replace(END_VARIABLE, "") + END_VARIABLE + afterInvocationArgs.replace(END_VARIABLE, "");
+            }
+        }
+
+        return snippetContent;
+    }
+
+    @Contract("null -> false")
```
Comment (SCWells72) [inline src/main/java/com/redhat/devtools/lsp4ij/client/features/LSPCompletionProposal.java:212]: I added `@Contract` annotations here and below because the IDE's nullability inspection was having trouble determining the relationships between the params and the results and was raising false inspection positives.

File: src/main/java/com/redhat/devtools/lsp4ij/server/definition/launching/UserDefinedCompletionFeature.java [hunk 5]
```diff
@@ -24,8 +25,15 @@ public class UserDefinedCompletionFeature extends LSPCompletionFeature {
 
     @Override
     public boolean useContextAwareSorting(@NotNull PsiFile file) {
-        UserDefinedLanguageServerDefinition serverDefinition = (UserDefinedLanguageServerDefinition) getClientFeatures().getServerDefinition();
+        ClientConfigurableLanguageServerDefinition serverDefinition = (ClientConfigurableLanguageServerDefinition) getClientFeatures().getServerDefinition();
```
Comment (SCWells72) [inline src/main/java/com/redhat/devtools/lsp4ij/server/definition/launching/UserDefinedCompletionFeature.java:28]: Previously it was deemed safe to blindly cast to `UserDefinedLanguageServerDefinition` so I'm doing the same here with `ClientConfigurableLanguageServerDefinition`. However, I would feel better checking the type before casting with this change. Thoughts? If you agree, I'll update all such usages to be more defensive.

File: src/main/resources/templates/clojure-lsp/clientSettings.json [hunk 6]
```diff
@@ -1,7 +1,8 @@
 {
   "caseSensitive": true,
   "completion": {
-    "useContextAwareSorting": true
+    "useContextAwareSorting": true,
+    "useTemplateForInvocationOnlySnippet": false
```
Comment (SCWells72) [inline src/main/resources/templates/clojure-lsp/clientSettings.json:5]: I updated all of the templates to disable editor templates for invocation-only snippets similar to how I enabled context-aware sorting in the templates.

File: src/test/java/com/redhat/devtools/lsp4ij/features/completion/CssCompletionClientConfigTest.java [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+ ******************************************************************************/
+
+package com.redhat.devtools.lsp4ij.features.completion;
+
+import com.redhat.devtools.lsp4ij.fixtures.LSPCompletionClientConfigFixtureTestCase;
+
+/**
+ * Tests code completion client configuration settings using a mock CSS language server.
+ */
+public class CssCompletionClientConfigTest extends LSPCompletionClientConfigFixtureTestCase {
```
Comment (SCWells72) [inline src/test/java/com/redhat/devtools/lsp4ij/features/completion/CssCompletionClientConfigTest.java:19]: The CSS test exercises whitespace-delimited argument list snippets as well as some complex snippets with variable values that are enumerated lists/choices.

File: src/test/java/com/redhat/devtools/lsp4ij/fixtures/LSPCompletionClientConfigFixtureTestCase.java [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+                                              @NotNull String fileBodyAfter,
+                                              @NotNull String mockTextCompletionJson,
+                                              @NotNull String mockCompletionItemResolveJson,
+                                              @Nullable Consumer<ClientConfigurationSettings> clientConfigCustomizer) {
+        MockLanguageServer.INSTANCE.setTimeToProceedQueries(200);
+
+        CompletionList mockCompletionList = JSONUtils.getLsp4jGson().fromJson(mockTextCompletionJson, CompletionList.class);
+        MockLanguageServer.INSTANCE.setCompletionList(mockCompletionList);
+
+        CompletionItem mockCompletionItem = JSONUtils.getLsp4jGson().fromJson(mockCompletionItemResolveJson, CompletionItem.class);
```
Comment (SCWells72) [inline src/test/java/com/redhat/devtools/lsp4ij/fixtures/LSPCompletionClientConfigFixtureTestCase.java:79]: I had to add the ability to specify the mock resolved completion item.

File: src/test/java/com/redhat/devtools/lsp4ij/fixtures/LSPCompletionClientConfigFixtureTestCase.java [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        try {
+            ContainerUtil.addAllNotNull(languageServers, LanguageServiceAccessor.getInstance(project)
+                    .getLanguageServers(file.getVirtualFile(), null, null)
+                    .get(5000, TimeUnit.MILLISECONDS));
+        } catch (Exception e) {
+            fail(e.getMessage());
+        }
+
+        // Configure the language server's completion client configuration
+        LanguageServerItem languageServer = ContainerUtil.getFirstItem(languageServers);
```
Comment (SCWells72) [inline src/test/java/com/redhat/devtools/lsp4ij/fixtures/LSPCompletionClientConfigFixtureTestCase.java:98]: Now we need to get the language server that's going to be used and set it up properly for client config testing.

File: src/test/java/com/redhat/devtools/lsp4ij/fixtures/LSPCompletionClientConfigFixtureTestCase.java [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+        // Use a configurable completion feature
+        LSPClientFeatures clientFeatures = languageServer.getClientFeatures();
+        clientFeatures.setCompletionFeature(new UserDefinedCompletionFeature());
+
+        // Enable completion item resolution
+        languageServer.getServerCapabilities().setCompletionProvider(new CompletionOptions(true, null));
+
+        // Update client configuration as required for this test scenario
+        if (clientConfigCustomizer != null) {
```
Comment (SCWells72) [inline src/test/java/com/redhat/devtools/lsp4ij/fixtures/LSPCompletionClientConfigFixtureTestCase.java:109]: And finally, get the client configuration from the server and apply any config changes for this test scenario.

File: src/test/java/com/redhat/devtools/lsp4ij/fixtures/LSPCompletionClientConfigFixtureTestCase.java [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+            LanguageServerDefinition languageServerDefinition = languageServer.getServerDefinition();
+            assertInstanceOf(languageServerDefinition, ClientConfigurableLanguageServerDefinition.class);
+            ClientConfigurableLanguageServerDefinition configurableLanguageServerDefinition = (ClientConfigurableLanguageServerDefinition) languageServerDefinition;
+            ClientConfigurationSettings clientConfiguration = configurableLanguageServerDefinition.getLanguageServerClientConfiguration();
+            assertNotNull(clientConfiguration);
+            clientConfigCustomizer.accept(clientConfiguration);
+        }
+
+        // Move to the offset at which completion should be triggered
```
Comment (SCWells72) [inline src/test/java/com/redhat/devtools/lsp4ij/fixtures/LSPCompletionClientConfigFixtureTestCase.java:118]: Hopefully the rest is pretty self-explanatory.

File: src/test/java/com/redhat/devtools/lsp4ij/mock/MockLanguageServer.java [hunk 12]
```diff
@@ -162,6 +162,10 @@ public void setCompletionList(CompletionList completionList) {
 		this.textDocumentService.setMockCompletionList(completionList);
 	}
 
+	public void setCompletionItem(CompletionItem completionItem) {
```
Comment (SCWells72) [inline src/test/java/com/redhat/devtools/lsp4ij/mock/MockLanguageServer.java:165]: Like I said above, need to be able to set the mock resolved completion item.

File: src/test/java/com/redhat/devtools/lsp4ij/mock/MockLanguageServerDefinition.java [hunk 13]
```diff
@@ -12,16 +12,21 @@
...
 import com.redhat.devtools.lsp4ij.server.definition.LanguageServerDefinition;
+import com.redhat.devtools.lsp4ij.server.definition.launching.ClientConfigurationSettings;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 /**
  * {@link LanguageServerDefinition} implementation to register the mock language server.
  */
-public class MockLanguageServerDefinition extends LanguageServerDefinition {
+public class MockLanguageServerDefinition extends LanguageServerDefinition implements ClientConfigurableLanguageServerDefinition {
```
Comment (SCWells72) [inline src/test/java/com/redhat/devtools/lsp4ij/mock/MockLanguageServerDefinition.java:24]: The mock language server is now client-configurable as well.

File: src/test/java/com/redhat/devtools/lsp4ij/mock/MockLanguageServerDefinition.java [hunk 14]
```diff
@@ -12,16 +12,21 @@
...
 
     private static final String SERVER_ID = "mock-server-id";
 
+    private final ClientConfigurationSettings clientConfigurationSettings = new ClientConfigurationSettings();
```
Comment (SCWells72) [inline src/test/java/com/redhat/devtools/lsp4ij/mock/MockLanguageServerDefinition.java:28]: And it just starts with a default client configuration.

File: src/main/java/com/redhat/devtools/lsp4ij/client/features/LSPCompletionProposal.java [hunk 15]
```diff
@@ -102,12 +109,22 @@ public void handleInsert(@NotNull InsertionContext context) {
...
             // Load the insert text to build:
             // - an IJ Template instance which will take care of replacement of placeholders
             // - the insert text without placeholders
-            template = SnippetTemplateFactory.createTemplate(snippetContent, context.getProject(), name -> getVariableValue(name), indentOptions);
+            template = SnippetTemplateFactory.createTemplate(
+                    snippetContent,
+                    context.getProject(),
+                    this::getVariableValue,
+                    indentOptions,
+                    useTemplateForInvocationOnlySnippet
```
Comment (SCWells72) [inline src/main/java/com/redhat/devtools/lsp4ij/client/features/LSPCompletionProposal.java:126]: We now pass this state to the snippet-to-template transform.

File: src/main/java/com/redhat/devtools/lsp4ij/features/completion/SnippetTemplateFactory.java [hunk 16]
```diff
@@ -33,15 +34,25 @@ public class SnippetTemplateFactory {
...
+                                                   LspSnippetIndentOptions indentOptions,
+                                                   boolean useTemplateForInvocationOnlySnippet) {
         Template template = TemplateManager.getInstance(project).createTemplate("", "");
         template.setInline(true);
-        LspSnippetParser parser = new LspSnippetParser(new SnippetTemplateLoader(template, variableResolver, indentOptions));
+        LspSnippetParser parser = new LspSnippetParser(new SnippetTemplateLoader(
+                template,
+                variableResolver,
+                indentOptions,
+                useTemplateForInvocationOnlySnippet
```
Comment (SCWells72) [inline src/main/java/com/redhat/devtools/lsp4ij/features/completion/SnippetTemplateFactory.java:54]: And propagate it to the template loader.

File: src/main/java/com/redhat/devtools/lsp4ij/features/completion/SnippetTemplateLoader.java [hunk 17]
```diff
@@ -32,43 +37,126 @@
 public class SnippetTemplateLoader extends AbstractLspSnippetHandler {
 
     private final Template template;
+    private final boolean useTemplateForInvocationOnlySnippet;
 
-    private final List<String> existingVariables;
+    private final List<SnippetTemplateSegment> templateSegments = new LinkedList<>();
```
Comment (SCWells72) [inline src/main/java/com/redhat/devtools/lsp4ij/features/completion/SnippetTemplateLoader.java:42]: Instead of building the template on-the-fly, we accumulate the exact same information that we'd need to build the template, and we do it in `endSnippet()` instead. That allows us to have full context of the resulting template segments so we can see whether/how it can be transformed/simplified.

File: src/main/java/com/redhat/devtools/lsp4ij/features/completion/SnippetTemplateLoader.java [hunk 18]
```diff
@@ -32,43 +37,126 @@
...
+                // If we find a variable not in an invocation, we can't simplify this template
+                else if ((templateSegment.isVariable() || templateSegment.isNamedVariable()) && !inInvocation) {
+                    hasVariablesOutsideInvocation = true;
+                    break;
+                }
+            }
 
+            // If we found an invocation and all variables were inside of it, use the simplified template segments
+            if (hasInvocation && !hasVariablesOutsideInvocation) {
+                effectiveTemplateSegments = simplifiedTemplateSegments;
```
Comment (SCWells72) [inline src/main/java/com/redhat/devtools/lsp4ij/features/completion/SnippetTemplateLoader.java:110]: If it looks like an invocation-only snippet, use the simplified segments below when building the actual template.

File: src/main/java/com/redhat/devtools/lsp4ij/features/completion/SnippetTemplateLoader.java [hunk 19]
```diff
@@ -32,43 +37,129 @@
...
     @Override
     public void endSnippet() {
+        List<SnippetTemplateSegment> effectiveTemplateSegments = templateSegments;
+
+        // If we should not use a template for an invocation-only snippet, see whether this fits the pattern
+        if (!useTemplateForInvocationOnlySnippet) {
+            boolean hasInvocation = false;
+            // While it's unlikely to happen in a typical invocation arg list scenario, this helps ensure we're only
+            // reacting to the contents of balanced parentheses
+            int nestedInvocationCount = 0;
```
Comment (SCWells72) [inline src/main/java/com/redhat/devtools/lsp4ij/features/completion/SnippetTemplateLoader.java:77]: This is an additional safeguard against the unlikely possibility of nested parens in segments. We'll only come out of this logic deciding to use the simplified text segments if everything occurred within balanced invocation parens.

File: src/main/java/com/redhat/devtools/lsp4ij/features/completion/SnippetTemplateLoader.java [hunk 20]
```diff
@@ -32,43 +37,129 @@
...
+    }
+
+    private static boolean isInvocationEnd(@NotNull String text) {
+        return StringUtil.trimLeading(text).startsWith(")");
     }
 
     @Override
     public void text(String text) {
-        template.addTextSegment(formatText(text));
+        templateSegments.add(SnippetTemplateSegment.textSegment(formatText(text)));
```
Comment (SCWells72) [inline src/main/java/com/redhat/devtools/lsp4ij/features/completion/SnippetTemplateLoader.java:154]: From here down, what you'll see is the change from building the template itself to accumulating what's needed to build the template in `endSnippet()`.

File: src/main/java/com/redhat/devtools/lsp4ij/features/completion/SnippetTemplateSegment.java [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+
+import com.intellij.codeInsight.template.Expression;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * This is a simple wrapper for a {@link com.intellij.codeInsight.template.Template} segment that allows deferred
+ * creation of the corresponding template with all segment information retained.
+ */
+class SnippetTemplateSegment {
```
Comment (SCWells72) [inline src/main/java/com/redhat/devtools/lsp4ij/features/completion/SnippetTemplateSegment.java:25]: And this is the way we accumulate that state. I've basically added factory methods for each of the types of template segment that are created above and simple ways to check which type of segment each is so it can be used safely/properly when building the final template.

File: src/test/java/com/redhat/devtools/lsp4ij/features/completion/SnippetTemplateLoaderTest.java [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+package com.redhat.devtools.lsp4ij.features.completion;
+
+import com.intellij.codeInsight.template.Template;
+import com.redhat.devtools.lsp4ij.fixtures.LSPCompletionFixtureTestCase;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Tests behavior of {@link SnippetTemplateLoader} via {@link SnippetTemplateFactory}.
+ */
+public class SnippetTemplateLoaderTest extends LSPCompletionFixtureTestCase {
```
Comment (SCWells72) [inline src/test/java/com/redhat/devtools/lsp4ij/features/completion/SnippetTemplateLoaderTest.java:24]: Direct tests for `SnippetTemplateLoader` to exercise common and corner-case scenarios.

**Comment by angelozerr** [comment]
It works like a charm, great job @SCWells72 !
